### PR TITLE
Ability to force a theme on Windows

### PR DESCRIPTION
**Pull request by VZout**
Added the `WindowBuilderExtWindows::with_theme` function to allow forcing a theme. In case you always want to have a dark theme no matter the system settings.

- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] ~~Created or updated an example program if it would help users understand this functionality~~
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by repi** [comment]
Any feedback about this or merging it in?

**Comment by chrisduerr** [comment]
I don't think there's any Windows maintainer capable of reviewing this.

File: src/platform_impl/windows/window_state.rs
Comment (filnet) [inline src/platform_impl/windows/window_state.rs:35]: would it make sense to get rid of is_dark_mode and use only theme ?
probably a breaking change.

Comment (VZout) [reply src/platform_impl/windows/window_state.rs:35]: So forcing a theme is added alongside the normal functionality of using the windows theme settings.
Also a window is not always use the preferred theme on old windows versions.

A approach could be extending the theme enum to be something like:
````rust
enum {
  OS_DETERMINED,
  FORCED_DARK,
  FORCED_LIGHT,
}
````
and `OS_DETERMINED` would be the default.


However this does not get rid of `is_dark` because using dark style can fail and it is used to check the OS preference.

**Review by msiglreith — APPROVED** [review]
Tested on win10:
- `None`: Correctly shows the theme selected in the OS
- `Some(theme)`: Correctly overrides the theme set by the user
- `None` + theme switch: Correctly switches to OS theme
- `Some(theme)` + theme swtich: Correctly keeps the forced theme

Good to go imo if there are no further code related concerns

File: CHANGELOG.md
Comment (kchibisov) [inline CHANGELOG.md:10]: The location is wrong, please rebase that PR properly.

I'd also word it like

```suggestion
- On Windows, add `WindowBuilderExtWindows::with_theme(Theme::)` to force specific theme.
```

File: FEATURES.md
Comment (kchibisov) [inline FEATURES.md:120]: I don't like `a`, since you're forcing some specific theme and not current one, etc. If you have other ideas on how to show the intent more clear, let me know.

```suggestion
* Theme the title bar according to Windows 10 Dark Mode setting or force preferred theme
```

File: src/platform/windows.rs
Comment (kchibisov) [inline src/platform/windows.rs:129]: Just by looking on that function, I feel like it'll make sense to accept `Option<Theme>`, meaning that providing `None` will use system's preferred `Theme`. That seems more ergonomic and can interact better with downstream configurations, since otherwise they should avoid calling that method, what do you think?

File: src/platform_impl/windows/mod.rs
Comment (kchibisov) [inline src/platform_impl/windows/mod.rs:24]: I don't really like the word `force` here, maybe we should use `prefer` and in some other places?

```suggestion
    pub preferred_theme: Option<Theme>,
```

**Review by kchibisov — CHANGES_REQUESTED** [review]
I don't have windows machine to test it, but I'd do some source level review, since I have some style questions.

File: src/platform_impl/windows/dark_mode.rs
Comment (kchibisov) [inline src/platform_impl/windows/dark_mode.rs:83]: Could you rewrite it with `let is_dark_mode = match preffered_theme`?

File: src/window.rs
Comment (kchibisov) [inline src/window.rs:868]: I don't like that change, and still prefer the use of `Option`, since we use `Theme` in `ThemeChanged` event, and so if the user  gets `System` it is confusing, since they don't know how to adapt to it. And even if winit can't send `System` users are still have to do something about it, so let's go back to `Option` approach to avoid that confusion.

Comment (kchibisov) [reply src/window.rs:868]: Note, you can internally map `None` to `System`, but you shouldn't expose that API to the user, since users don't have a way to handle `System` variant.

File: src/platform_impl/windows/event_loop.rs
Comment (kchibisov) [inline src/platform_impl/windows/event_loop.rs:1881]: Could you 'inline' changed thing?

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: src/platform_impl/windows/event_loop.rs
Comment (VZout) [reply src/platform_impl/windows/event_loop.rs]: Done! Also implemented your other feedback.

File: src/platform_impl/windows/dark_mode.rs
Comment (filnet) [inline src/platform_impl/windows/dark_mode.rs:79]: do you really need to specify the bool type ?

File: src/platform_impl/windows/event_loop.rs
Comment (filnet) [inline src/platform_impl/windows/event_loop.rs:1874]: not sure you need to clone (it's an enum)

File: src/platform_impl/windows/window.rs
Comment (filnet) [inline src/platform_impl/windows/window.rs:729]: is clone() needed ?

File: src/platform/windows.rs
Comment (kchibisov) [inline src/platform/windows.rs:129]:
```suggestion
    /// Forces a theme or uses the system settings if `None` was provided.
```

File: src/platform_impl/windows/window_state.rs
Comment (kchibisov) [inline src/platform_impl/windows/window_state.rs:34]: I feel like for the more code simplicity we can use `Theme` instead of `bool` here, so `try_dark_mode` will also return a `Theme`, and then you'd have less mapping to `bool -> Theme` , issuing a new `ThemeChanged` when `current_theme != new_theme`?

**Review by kchibisov — CHANGES_REQUESTED** [review]
That looks better and I feel like we're getting closer! Thanks for your work so  far.

File: src/platform_impl/windows/window_state.rs
Comment (VZout) [reply src/platform_impl/windows/window_state.rs]: Yeah I like this changed. Implemented it!

File: src/platform_impl/windows/dark_mode.rs
Comment (kchibisov) [inline src/platform_impl/windows/dark_mode.rs:76]: The comment isn't not correct anymore, we now return `Theme`.

File: src/platform_impl/windows/dark_mode.rs
Comment (kchibisov) [inline src/platform_impl/windows/dark_mode.rs:77]: I'd actually rename this function, since it it's not about dark mode anymore, it's just about trying to get a theme, however when I see try, I'm expecting `Result`, so I'm not sure if this is a good name. Maybe `pick_theme` or something like that, if you have better ideas let me know.

File: CHANGELOG.md
Comment (kchibisov) [inline CHANGELOG.md:4]: Should append into the changelog, and not insert between. Also, there's no need to provide argument in changelog, just a name.

```suggestion
- On Windows, add `WindowBuilderExtWindows::with_theme` to set a preferred theme.
```

**Review by kchibisov — CHANGES_REQUESTED** [review]
You should address compilation warnings, besides that it should be good to go.

File: src/platform_impl/windows/dark_mode.rs
Comment (VZout) [reply src/platform_impl/windows/dark_mode.rs:77]: if your preferred theme is dark it could theoretically fail on old windows versions. In that case it would return `Theme::Light`.
So I think `try` should be included in the name. `try_select_theme`, `try_pick_theme` or `try_theme` 🤷 

I don't think it should necessarily return a result. I could change it to a result but it would make the calling code less nice IMO. (It would return a result, if the result is ok the current theme is the preferred theme if the result is a error than the theme is light. Which works but the caller of `try_dark_mode` needs to understand that the standard theme is light so I prefer to hide it away into the `try_` function)

File: CHANGELOG.md
Comment (VZout) [reply CHANGELOG.md]: 👍 commited

File: src/platform_impl/windows/dark_mode.rs
Comment (VZout) [reply src/platform_impl/windows/dark_mode.rs]: fixed

**Review by kchibisov — APPROVED** [review]
Needs rebase and should be good to go.

File: src/platform/windows.rs
Comment (msiglreith) [inline src/platform/windows.rs:84]: Documentation doesn't match anymore

File: src/platform/windows.rs
Comment (msiglreith) [inline src/platform/windows.rs:85]: This seems like a breaking change, so it should be listed in the changelog

Comment (VZout) [reply src/platform/windows.rs:85]: Changed the comment

File: src/platform/windows.rs
Comment (VZout) [reply src/platform/windows.rs]: Changed the comment

File: CHANGELOG.md
Comment (msiglreith) [inline CHANGELOG.md:3]:
```suggestion
- **Breaking:** On Windows, Renamed `WindowBuilderExtWindows`'s `is_dark_mode` to `theme`.
```
(see https://github.com/rust-windowing/winit/blob/1ef38f04127d3399b792aefcbcb233cf390ab5d7/CHANGELOG.md)

**Pull request merged**
